### PR TITLE
[dtensor] implement distributed topk operator

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126713
* #126712
* __->__ #126711
* #126710

as titled. Implemented the topk operator in DTensor.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: test/distributed/_tensor/test_math_ops.py [hunk 1]
```diff
@@ -394,6 +394,26 @@ def _replicate_fn(name, module, device_mesh):
 
             self.assertEqual(x_local.grad, x_dist.grad.full_tensor())
 
+    @with_comms
+    def test_topk(self):
+        device_mesh = self.build_device_mesh()
+        placement_combs = [Shard(0), Shard(1), Shard(2), Replicate()]
+
+        comm_mode = CommDebugMode()
```
Comment (wz337) [inline test/distributed/_tensor/test_math_ops.py:402]: nit: looks like comm_mode is not used in the test?

Comment (wanchaol) [reply test/distributed/_tensor/test_math_ops.py:402]: good catch! updated

**Review by wz337 — APPROVED** [review]
LGTM! But will probably need to rebase on top Tianyu's change. https://github.com/pytorch/pytorch/pull/126359

File: torch/distributed/_tensor/ops/math_ops.py [hunk 2]
```diff
@@ -954,26 +981,55 @@ def layer_norm_bwd_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy
...
-            new_placements.append(p)  # keep the placement
-    return tuple(new_placements)
+@register_op_strategy(
+    [aten.topk.default],
+    schema_info=RuntimeSchemaInfo(2),
+)
+def topk_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+    input_strategy = op_schema.args_schema[0]
+    k = op_schema.args_schema[1]
+    input_shape = input_strategy.output_shape
```
Comment (wz337) [inline torch/distributed/_tensor/ops/math_ops.py:991]: output_shape -> shape after rebase.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 3]
```diff
@@ -954,26 +981,55 @@ def layer_norm_bwd_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy
...
+    topk_dim = op_schema.args_schema[2] if len(op_schema.args_schema) > 2 else -1
+    topk_dim = normalize_dim(topk_dim, input_strategy.output_ndim)
```
Comment (wz337) [inline torch/distributed/_tensor/ops/math_ops.py:993]: output_ndim -> ndim after rebase.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 4]
```diff
@@ -954,26 +981,55 @@ def layer_norm_bwd_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy
...
+    for mesh_dim in range(mesh.ndim):
+        single_mesh_dim_strategies = []
+
+        # two outputs (values, indices), 1 input
+        # replicate always works
+        all_replicate: List[Placement] = [Replicate()] * 3
+        single_mesh_dim_strategies.append(all_replicate)
+
+        # every dim except topk dim should work
+        for dim in range(input_strategy.output_ndim):
```
Comment (wz337) [inline torch/distributed/_tensor/ops/math_ops.py:1006]: ditto

**Pull request closed without merging**